### PR TITLE
Refactor plugin, component class, component APIs

### PR DESCRIPTION
**Pull request by eepp**
I made sure to put as much detail as possible in the commit messages to make those changes as clear as possible.

Note that 218e98ed03e51c9ce9098e4a6a1c0ffbf5169095 is further modified by 4f35d2a21c8c65e922761516d2e277c88a72b043, but the spirit remains the same. Feel free to combine them if you feel that they belong together.

### CODE REVIEW

File: include/babeltrace/plugin/plugin-dev.h [hunk 1]
```diff
@@ -31,90 +31,376 @@
...
+	BT_PLUGIN_DESCRIPTOR_ATTRIBUTE_TYPE_DESCRIPTION		= 4,
+};
+
+/* Plugin attribute (internal use) */
+struct __bt_plugin_descriptor_attribute {
+	/* Plugin descriptor to which to associate this attribute */
+	const struct __bt_plugin_descriptor *plugin_descriptor;
+
+	/* Attribute's type */
+	enum __bt_plugin_descriptor_attribute_type type;
```
Comment (compudj) [inline include/babeltrace/plugin/plugin-dev.h]: move "type" after type_name field.

Comment (compudj) [inline include/babeltrace/plugin/plugin-dev.h]: for the other struct below too. I'm tempted to keep the type close to the union it helps selecting. And on the plus side, it allows the pointers to be aligned in memory (as a side-effect).

File: include/babeltrace/plugin/plugin-dev.h [hunk 2]
```diff
@@ -31,90 +31,376 @@
...
+ * _comp_class_id: Component class ID (C identifier).
+ * _type:          Component class type (enum bt_component_type).
+ * _attr_name:     Name of the attribute (C identifier).
+ * _attr_type:     Type of the attribute
+ *                 (enum __bt_plugin_descriptor_attribute_type).
+ * _x:             Value.
+ */
+#define __BT_PLUGIN_COMPONENT_CLASS_DESCRIPTOR_ATTRIBUTE(_attr_name, _attr_type, _id, _comp_class_id, _type, _x) \
+	static struct __bt_plugin_component_class_descriptor_attribute __bt_plugin_component_class_descriptor_attribute_##_id##_##_comp_class_id##_##_type##_##_attr_name = { \
+		.comp_class_descriptor = &__bt_plugin_component_class_descriptor_##_id##_##_comp_class_id##_##_type, \
```
Comment (compudj) [inline include/babeltrace/plugin/plugin-dev.h]: did we want to add documentation in this header to cover cases where a linking error happens due to missing component or filter symbol ?

File: plugins/ctf/fs/fs.c [hunk 3]
```diff
@@ -723,10 +723,11 @@ struct ctf_fs_component *ctf_fs_create(struct bt_value *params)
 
 BT_HIDDEN
 enum bt_component_status ctf_fs_init(struct bt_component *source,
-		struct bt_value *params)
+		struct bt_value *params, void *init_method_data)
 {
 	struct ctf_fs_component *ctf_fs;
 	enum bt_component_status ret = BT_COMPONENT_STATUS_OK;
+	(void) init_method_data;
```
Comment (compudj) [inline plugins/ctf/fs/fs.c]: is it a __used or __unused attribute that you want here on the func argument instead ?

Comment (compudj) [inline plugins/ctf/fs/fs.c]: same comment applies further below.

File: include/babeltrace/component/component-class.h [hunk 4]
```diff
@@ -50,7 +50,8 @@ enum bt_component_class_type {
 };
 
 typedef enum bt_component_status (*bt_component_class_init_method)(
-		struct bt_component *component, struct bt_value *params);
+		struct bt_component *component, struct bt_value *params,
+		void *init_method_data);
```
Comment (compudj) [inline include/babeltrace/component/component-class.h]: so we add a private pointer, but where is it kept ? and how is it passed to the callbacks ?

File: plugins/ctf/fs/fs.c (hunk 3 above)
Comment (eepp) [inline plugins/ctf/fs/fs.c]: Well casting to void is just another, portable way to achieve this? What do you suggest?

**Comment by compudj** [comment]
----- On Jan 26, 2017, at 11:48 AM, Philippe Proulx <notifications@github.com> wrote: 

> @eepp commented on this pull request.

> In [ https://github.com/jgalar/babeltrace/pull/18 | plugins/ctf/fs/fs.c ] :
> >  {
>  	struct ctf_fs_component *ctf_fs;
>  	enum bt_component_status ret = BT_COMPONENT_STATUS_OK;
> +	(void) init_method_data;

> Well casting to void is just another, portable way to achieve this? What do you
> suggest?
int foo ( __attribute__ (( unused )) int bar ) { return 0 ; } 
https://gcc.gnu.org/onlinedocs/gcc-3.2/gcc/Variable-Attributes.html 

Thanks, 

Mathieu 

-- 
Mathieu Desnoyers 
EfficiOS Inc. 
http://www.efficios.com

File: include/babeltrace/component/component-class.h (hunk 4 above)
Comment (eepp) [inline include/babeltrace/component/component-class.h]: It's some custom user data for the init method exclusively. The init method has the responsability of doing something with it. There must be a contract between the caller of `bt_component_create_with_init_method_data()` and the init method that's eventually called: either they are both implemented by the same developer, or the caller knows what to pass to the specific init method and what happens with the ownership (transfered, reference count increments, belongs to some GC manager anyway, etc.).

`bt_component_create_with_init_method_data()` is just a utility which is reserved for very specific use cases. It won't be used by `babeltrace` (the converter), and thus if you write a plugin for this, the `init_method_data` parameter will always be set to `NULL` in your init method callbacks.

**Comment by eepp** [comment]
@jgalar BTW, documentation for all this (and for existing stuff which is not or not accurately documented) is postponed to after RC1.

**Comment by eepp** [comment]
> int foo ( __attribute__ (( unused )) int bar ) { return 0 ; }
> https://gcc.gnu.org/onlinedocs/gcc-3.2/gcc/Variable-Attributes.html

I'll let @jgalar take a decision here.

**Comment by compudj** [comment]
----- On Jan 26, 2017, at 11:52 AM, Philippe Proulx <notifications@github.com> wrote: 

> @eepp commented on this pull request.

> In [ https://github.com/jgalar/babeltrace/pull/18 |
> include/babeltrace/component/component-class.h ] :
> > @@ -50,7 +50,8 @@ enum bt_component_class_type {
>  };

>  typedef enum bt_component_status (*bt_component_class_init_method)(
> -		struct bt_component *component, struct bt_value *params);
> +		struct bt_component *component, struct bt_value *params,
> +		void *init_method_data);

> It's some custom user data for the init method exclusively. The init method has
> the responsability of doing something with it. There must be a contract between
> the caller of bt_component_create_with_init_method_data() and the init method
> that's eventually called: either they are both implemented by the same
> developer, or the caller knows what to pass to the specific init method and
> what happens with the ownership (transfered, reference count increments,
> belongs to some GC manager anyway, etc.).

> bt_component_create_with_init_method_data() is just a utility which is reserved
> for very specific use cases. It won't be used by babeltrace (the converter),
> and thus if you write a plugin for this, the init_method_data parameter will
> always be set to NULL in your init method callbacks.
No, I actually intend to use it for the lttng-live source plugin. 

Mathieu 

-- 
Mathieu Desnoyers 
EfficiOS Inc. 
http://www.efficios.com

**Comment by jgalar** [comment]
On 26 January 2017 at 11:55, Philippe Proulx <notifications@github.com>
wrote:

> @jgalar <https://github.com/jgalar> BTW, documentation for all this (and
> for existing stuff which is not or not accurately documented) is postponed
> to after RC1.
>

Do you keep a list somewhere of the documentation to be updated? I'm not
sure we'll have the time to comprehensively re-review all of it before the
release.

**Comment by eepp** [comment]
> Do you keep a list somewhere of the documentation to be updated?

Yes, it's basically any public header file without a Doxygen module (`@defgroup`).

File: plugins/ctf/fs/fs.c (hunk 3 above)
Comment (jgalar) [inline plugins/ctf/fs/fs.c]: We make use of various GNU extensions in the project, so I'm not against using `__attribute__((__unused__))` considering both GCC and clang support it.

However, please define a macro in compiler.h as is done elsewhere to isolate GNU-ish uglyness (e.g. see `BT_HIDDEN`).

Comment (eepp) [inline plugins/ctf/fs/fs.c]: It's actually in `babeltrace-internal.h` and I can add it there (`BT_UNUSED`) if that's okay.

File: tests/lib/test-plugin-plugins/minimal.c [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+static enum bt_plugin_status plugin_exit(void)
+{
+    test_plugin_exit_called = 1;
+
+	return BT_PLUGIN_STATUS_OK;
+}
+
+BT_PLUGIN_NAME("test-minimal");
+BT_PLUGIN_DESCRIPTION("Minimal Babeltrace plugin with no component classes");
+BT_PLUGIN_AUTHOR("Janine Sutto");
```
Comment (jgalar) [inline tests/lib/test-plugin-plugins/minimal.c]: Hahaha

Comment (simark) [inline tests/lib/test-plugin-plugins/minimal.c]: Janine Sutto

**Comment by jgalar** [comment]
Merged, thanks!

**Pull request closed without merging**